File: code/ch16/16.3.1.subsets.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        dfs(nums, 0, new ArrayList<>());
        return ans;
    }
    public void dfs(int[] nums, int idx, List<Integer> path) {
        if (idx == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }
        path.add(nums[idx]);
        dfs(nums, idx + 1, path);
        path.remove(path.size() - 1);
        dfs(nums, idx + 1, path);
    }
}

File: code/ch16/16.3.1.subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)

        def dfs(idx: int, path: List[int]):
            if idx == n:
                ans.append(path[:])
                return
            path.append(nums[idx])
            dfs(idx + 1, path)
            path.pop()
            dfs(idx + 1, path)

        dfs(0, [])
        return ans
